File: backend/schemas/crosspost.py
"""Cross-posting schemas."""

from __future__ import annotations

from pydantic import BaseModel, Field


class SocialAccountCreate(BaseModel):
    """Request to connect a social media account."""

    platform: str = Field(min_length=1, description="Platform name, e.g. 'bluesky' or 'mastodon'")
    account_name: str | None = Field(default=None, description="Display name for the account")
    credentials: dict[str, str] = Field(
        description="Platform-specific credentials (stored as JSON)"
    )


class SocialAccountResponse(BaseModel):
    """Response for a connected social account."""

    id: int
    platform: str
    account_name: str | None = None
    created_at: str


class CrossPostRequest(BaseModel):
    """Request to cross-post a blog post."""

    post_path: str = Field(min_length=1, description="Relative file path of the post to cross-post")
    platforms: list[str] = Field(min_length=1, description="List of platform names to post to")
    custom_text: str | None = Field(
        default=None, description="Optional custom text to post instead of auto-generated content"
    )


class CrossPostResponse(BaseModel):
    """Response for a single cross-post result."""

    id: int
    post_path: str
    platform: str
    platform_id: str | None = None
    status: str
    posted_at: str | None = None
    error: str | None = None


class CrossPostHistoryResponse(BaseModel):
    """Response containing cross-post history for a post."""

    items: list[CrossPostResponse]


class BlueskyAuthorizeRequest(BaseModel):
    """Request to start Bluesky OAuth flow."""

    handle: str = Field(min_length=1, description="Bluesky handle, e.g. 'alice.bsky.social'")


class BlueskyAuthorizeResponse(BaseModel):
    """Response with authorization URL for Bluesky OAuth."""

    authorization_url: str


class MastodonAuthorizeRequest(BaseModel):
    """Request to start Mastodon OAuth flow."""

    instance_url: str = Field(
        min_length=1, description="Mastodon instance URL, e.g. 'https://mastodon.social'"
    )


class MastodonAuthorizeResponse(BaseModel):
    """Response with authorization URL for Mastodon OAuth."""

    authorization_url: str

File: frontend/src/api/crosspost.ts
import api from './client'

export interface SocialAccount {
  id: number
  platform: string
  account_name: string | null
  created_at: string
}

export interface CrossPostResult {
  id: number
  post_path: string
  platform: string
  platform_id: string | null
  status: string
  posted_at: string | null
  error: string | null
}

export interface CrossPostHistory {
  items: CrossPostResult[]
}

export async function fetchSocialAccounts(): Promise<SocialAccount[]> {
  return api.get('crosspost/accounts').json<SocialAccount[]>()
}

export async function deleteSocialAccount(accountId: number): Promise<void> {
  await api.delete(`crosspost/accounts/${accountId}`)
}

export async function authorizeBluesky(
  handle: string,
): Promise<{ authorization_url: string }> {
  return api
    .post('crosspost/bluesky/authorize', { json: { handle } })
    .json<{ authorization_url: string }>()
}

export async function authorizeMastodon(
  instanceUrl: string,
): Promise<{ authorization_url: string }> {
  return api
    .post('crosspost/mastodon/authorize', { json: { instance_url: instanceUrl } })
    .json<{ authorization_url: string }>()
}

export async function authorizeX(): Promise<{ authorization_url: string }> {
  return api
    .post('crosspost/x/authorize')
    .json<{ authorization_url: string }>()
}

export async function authorizeFacebook(): Promise<{ authorization_url: string }> {
  return api
    .post('crosspost/facebook/authorize')
    .json<{ authorization_url: string }>()
}

export interface FacebookPage {
  id: string
  name: string
  access_token: string
}

export async function selectFacebookPage(
  state: string,
  pageId: string,
): Promise<{ account_name: string }> {
  return api
    .post('crosspost/facebook/select-page', {
      json: { state, page_id: pageId },
    })
    .json<{ account_name: string }>()
}

export async function crossPost(
  postPath: string,
  platforms: string[],
  customText?: string,
): Promise<CrossPostResult[]> {
  return api
    .post('crosspost/post', {
      json: { post_path: postPath, platforms, custom_text: customText ?? null },
    })
    .json<CrossPostResult[]>()
}

export async function fetchCrossPostHistory(
  postPath: string,
): Promise<CrossPostHistory> {
  return api.get(`crosspost/history/${postPath}`).json<CrossPostHistory>()
}
